src/java/org/apache/cassandra/index/sai/disk/v3/BlockTermsReader.java

```diff
@@ -0,0 +1,870 @@
+                                 QueryEventListener.BKDIndexEventListener listener,
+                                 QueryContext context) throws IOException
+    {
+        final PointValues.Relation relation = visitor.compare(new BytesRef(meta.minTerm), new BytesRef(meta.maxTerm));
+
+        if (relation == PointValues.Relation.CELL_OUTSIDE_QUERY)
+        {
+            listener.onIntersectionEarlyExit();
+            return null;
+        }
+
+        final LongArray postingBlockFPs = new LongArray.DeferredLongArray(blockPostingOffsetsReader::open);
+        final LongArray orderMapFPs = new LongArray.DeferredLongArray(blockOrderMapFPReader::open);
+
+        listener.onSegmentHit();
+
+        // it's important to create the input streams once per query as
+        // there is a significant performance cost atm
+        final IndexInputReader treeIndexInput = IndexInputReader.create(termsIndexHandle);
+        final IndexInputReader postingsInput = IndexInputReader.create(postingsHandle);
+        final IndexInputReader upperPostingsInput = IndexInputReader.create(upperPostingsHandle);
+
+        try (final BinaryTree.Reader treeReader = new BinaryTree.Reader(meta.numPostingBlocks,
+                                                                        meta.pointCount,
+                                                                        meta.postingsBlockSize,
+                                                                        treeIndexInput))
+        {
+            final Intersection intersection = relation == PointValues.Relation.CELL_INSIDE_QUERY ?
+                                              new Intersection(postingsInput,
+                                                               upperPostingsInput,
+                                                               treeReader,
+                                                               postingBlockFPs,
+                                                               orderMapFPs,
+                                                               context,
+                                                               listener) :
+                                              new FilteringIntersection(postingsInput,
+                                                                        upperPostingsInput,
+                                                                        treeReader,
+                                                                        visitor,
+                                                                        postingBlockFPs,
+                                                                        orderMapFPs,
+                                                                        context,
+                                                                        listener);
+
+            return intersection.execute();
+        }
+    }
+
+    private class Intersection
```

Comment:
What does it intersect with what? Takes quite many parameters, this is not obvious without reading the details of the implementation.

Reply:
Class structure and name is copied from the kdtree for easier comparison of the two.  Comparing the kdtree implementation was used to ensure the metrics were correct.  The kdtree is being refactored as part of CEP-7 cc @mike-tr-adamson so I believe it's best to change the kdtree similar classes later.